src/remote/publication.rs

```diff
@@ -0,0 +1,1205 @@
+}
+
+impl fmt::Display for ReportErrorCode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ReportErrorCode::XmlError => write!(f, "xml_error"),
+            ReportErrorCode::PermissionFailure => write!(f, "permission_failure"),
+            ReportErrorCode::BadCmsSignature => write!(f, "bad_cms_signature"),
+            ReportErrorCode::ObjectAlreadyPresent => write!(f, "object_already_present"),
+            ReportErrorCode::NoObjectPresent => write!(f, "no_object_present"),
+            ReportErrorCode::NoObjectMatchingHash => write!(f, "no_object_matching_hash"),
+            ReportErrorCode::ConsistencyProblem => write!(f, "consistency_problem"),
+            ReportErrorCode::OtherError => write!(f, "other_error"),
+        }
+    }
+}
+
+impl FromStr for ReportErrorCode {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "xml_error" => Ok(ReportErrorCode::XmlError),
+            "permission_failure" => Ok(ReportErrorCode::PermissionFailure),
+            "bad_cms_signature" => Ok(ReportErrorCode::BadCmsSignature),
+            "object_already_present" => Ok(ReportErrorCode::ObjectAlreadyPresent),
+            "no_object_present" => Ok(ReportErrorCode::NoObjectPresent),
+            "no_object_matching_hash" => Ok(ReportErrorCode::NoObjectMatchingHash),
+            "consistency_problem" => Ok(ReportErrorCode::ConsistencyProblem),
+            "other_error" => Ok(ReportErrorCode::OtherError),
+            _ => Err(Error::InvalidErrorCode(s.to_string())),
+        }
+    }
+}
+
+impl ReportErrorCode {
+    /// Provides default texts for error codes (taken from RFC).
+    fn to_text(&self) -> &str {
+        match self {
+            ReportErrorCode::XmlError => "Encountered an XML problem.",
+            ReportErrorCode::PermissionFailure => "Client does not have permission to update this URI.",
+            ReportErrorCode::BadCmsSignature => "Encountered bad CMS signature.",
+            ReportErrorCode::ObjectAlreadyPresent => "An object is already present at this URI, yet a \"hash\" attribute was not specified.",
+            ReportErrorCode::NoObjectPresent => "There is no object present at this URI, yet a \"hash\" attribute was specified.",
+            ReportErrorCode::NoObjectMatchingHash => "The \"hash\" attribute supplied does not match the \"hash\" attribute of the object at this URI.",
+            ReportErrorCode::ConsistencyProblem => "Server detected an update that looks like it will cause a consistency problem (e.g., an object was deleted, but the manifest was not updated).",
+            ReportErrorCode::OtherError => "Found some other issue."
+        }
+    }
+}
+
+
+//------------ Base64 --------------------------------------------------------
+
+/// This type contains a base64 encoded structure. The publication protocol
+/// deals with objects in their base64 encoded form.
+#[derive(Clone, Debug, Eq, PartialEq)]
```

Comment:
Would it make sense to keep the data in plain form and only do the base 64 encoding and decoding when and if necessary?

Reply:
It would, but.. I use this type in Krill and it became very slow when an earlier implementation stored `Bytes` and did the encoding/decoding on de/serializing. Especially when the Publication Server was starting up.

So, then I decided to move the work to when the actual bytes are needed, which is less frequently - for Krill at least.

I can add a remark to this effect - we can also change it later or let Krill use another type instead if we have to.

Reply:
Nah, it’s fine. Was just a thought – but if you are using it this way already for storage, then changing it will be unnecessary pain.